src/passes/Heap2Local.cpp

```diff
@@ -813,11 +793,12 @@ struct Struct2Local : PostWalker<Struct2Local> {
       refinalize = true;
     }
     Expression* value = builder.makeLocalGet(localIndexes[curr->index], type);
-    if (field.isPacked() && curr->signed_) {
-      // The value in the local is the masked unsigned value, which we must
-      // sign-extend.
-      value = Bits::makeSignExt(value, field.getByteSize(), wasm);
-    }
+    // Note that in theory we could try to do better here than to fix up the
+    // packing and signedness on gets: we could truncate on sets. That would be
+    // more efficient if all gets are unsigned, as gets outnumber sets in
+    // general. However, signed gets make that more complicated, so leave this
+    // for other opts to handle.
```

Comment:
How do signed gets make that more complicated? If you truncate the value on the set and sign extend it again on the get, you'll still get the correct value back. Can't we leave the redundant truncation and sign extension to be cleaned up by other ops instead?

Reply:
It's more complicated in that it's hard to see what is best here. That is, correctness is easy but we'd need to count how many gets are signed and how many are unsigned, to know which model is more efficient. So it is simplest to do the simple thing (fix up reads in one place), and leave optimizing that further to another pass.

(Atm nothing optimizes this further, but we could do it - basically count the sign-ext operations on local.gets etc. and maybe decide to shift from fixups on gets to sets or vice versa. We actually do something similar for linear memory loads and stores, but not locals.)

Reply:
Or, to put it another way, this PR simplifies Heap2Local code while also making it more efficient in at least one case (the case where there are many signed reads and many writes: we did the worst possible thing in that case before, which was to do a fixup on both those reads and those writes).

Reply:
Makes sense 👍